controllers/pipelinerun_controller.go

```diff
@@ -0,0 +1,215 @@
+	)
+
+	// search for Builds with Pipeline triggers matching current ObjectRef criteria
+	buildsToBeIssued := r.buildInventory.SearchForObjectRef(v1alpha1.PipelineTrigger, objectRef)
+	if len(buildsToBeIssued) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	buildNames := inventory.ExtractBuildNames(buildsToBeIssued...)
+	logger.V(0).Info("Build names in the Inventory matching criteria", "build-names", buildNames)
+
+	// during pipeline re-run a new PipelineRun is issued based on a existing object copying over all
+	// the elements, including annotations. To allow re-runs we annotate the current object's name
+	// and only check previously triggered builds when the name matches
+	var triggeredBuilds = []filter.TriggeredBuild{}
+	if filter.PipelineRunAnnotatedNameMatchesObject(&pipelineRun) {
+		// extracting existing triggered-builds from the annotation, information needed to detect if
+		// the BuildRuns have already beeing issued for the PipelineRun
+		triggeredBuilds, err = filter.PipelineRunExtractTriggeredBuildsSlice(&pipelineRun)
+		if err != nil {
+			logger.V(0).Error(err, "parsing triggered-builds annotation")
+			// in case of errors an empty slice takes place, may incur the side effect of issuing
+			// duplicated BuildRuns
+			triggeredBuilds = []filter.TriggeredBuild{}
+		}
+
+		// filtering out the instances that have already been processed, the annotation extracted
+		// shows  which build names and the objectRef employed
+		if filter.TriggereBuildsContainsObjectRef(triggeredBuilds, buildNames, objectRef) {
+			logger.V(0).Info("BuildRuns for PipelineRun have already been issued!")
+			return ctrl.Result{}, nil
+		}
+	} else {
+		logger.V(0).Info("PipelineRun annotated name does not match current object!")
+	}
+	logger.V(0).Info("PipelineRun previously triggered builds", "triggered-builds", triggeredBuilds)
+
+	// firing the BuildRun instances for the informed Builds
+	buildRunsIssued, err := r.issueBuildRunsForPipelineRun(ctx, &pipelineRun, buildNames)
+	if err != nil {
+		logger.V(0).Error(err, "trying to issue BuildRun instances")
+		return ctrl.Result{}, err
+	}
+	logger.V(0).Info("BuildRuns issued", "buildruns", buildRunsIssued)
+
+	// updating annotation appending the current state which triggered BuildRuns instances, this
+	// annotation is later on checked to skip the conditions that already triggered builds
+	if err = filter.PipelineRunAppendTriggeredBuildsAnnotation(
+		&pipelineRun,
+		triggeredBuilds,
+		buildNames,
+		objectRef,
+	); err != nil {
+		logger.V(0).Error(err, "trying to updated triggered-builds annotation")
+		return ctrl.Result{}, err
+	}
+
+	// updating label registering all BuildRuns issued
+	filter.AppendIssuedBuildRunsLabel(&pipelineRun, buildRunsIssued)
+	// annotating object's current name
+	filter.PipelineRunAnnotateName(&pipelineRun)
+
+	// patching the PipelineRun to reflect labels and annotations needed on the object
+	if err = r.Client.Patch(ctx, &pipelineRun, client.Merge); err != nil {
```

Comment:
I did not find out how this patch works. Browsing through the code it reads is if it takes the state of the object and patches it fully which means that it overwrites changes.

To verify one would need to go something like this:

```
obj1 = client.get("id")
obj2 = client.get("id")

obj1.field1 = "newValue1"
client.update(obj1)

obj2.field2 = "newValue2"
client.patch(obj2, client.Merge)
```

What will be the value of field1 ?

Should we maybe do this:

```
obj = client.get("id")
originalObj = obj.DeepClone()

obj.field = "newValue"

client.patch(obj, client.MergeFrom(originalObj))
```

Reply:
Indeed, let me adopt your suggestion. Please consider the latest changes.